### DISCUSSION TITLE
Wrong unsoundness?

### DISCUSSION THREAD

**Discussion by AntoineD** [python/typing#1331] [Q&A]
The following reproducer code returns the `__init__` method of a class:
```python
from typing import Any
def f(cls: type) -> Any:
    return cls.__init__
```

but mypy considers that the `__init__`method was accessed from an instance:
`main.py:4: error: Accessing "__init__" on an instance is unsound, since instance.__init__ could be from an incompatible subclass  [misc]`

Where is the problem?

**Comment by Gobot1234** [comment]
The error isn't incorrect, cls is technically an instance of type and so accessing __init__ which isn't checked for violations of the LSP is technically unsafe

**Reply by AntoineD** [reply]
I am not sure to understand.

Also, why is the following not raising an error?
```python
def f(cls: type) -> Any:
    return cls.__new__
```

**Reply by erictraut** [reply]
I don't think this should be an error. It looks to me like a mypy bug. I recommend filing a bug in the mypy issue tracker. 

FWIW, this type checks without errors in pyright.

**Reply by AntoineD** [reply]
@srittau Should it be converted back to an issue?

**Reply by hauntsaninja** [reply]
Like Gobot1234 alludes to, the error exists to catch the following issue:
```
class A:
    def __init__(self, arg1: int) -> None: ...

class B(A):
    def __init__(self, arg1: int, arg2: str) -> None: ...

def foo(a: A):
    a.__init__(1)

foo(B(1, "one"))  # oops, TypeError
```
(Note that pyright has a false negative in the above cc @erictraut )

I agree that the error isn't particularly useful in your case, but your code is so reduced that I'm having a hard time understanding the use case / you may as well type ignore.

In most real life use cases that this comes up, users can get it to type check by using the class instead of the instance. That is, something like `A.__init__(a, 1)` in the above instead of `a.__init__(1)`

**Reply by AntoineD** [reply]
@hauntsaninja Thank you for the clear explanation.
I can see the potential issue, but in my case this cannot be an actual issue since only the `__init__` method matters and not the result of its call (which is never performed). My [use case](https://github.com/AntoineD/docstring-inheritance/blob/62eb910471bf68e2edd3ae547687cf45450893a0/src/docstring_inheritance/class_docstrings_inheritor.py#L94) comes from a tool to process docstrings.

**Comment by UtkarshaGupte** [comment]
Is this a bug that mypy is tracking, if yes, can anyone share the issue link? I am also seeing the same issue. Can anyone enlighten me ?

**Reply by Hnasar** [reply]
@UtkarshaGupte I don't think there's a bug, would you like to file one?
